docs/dynamic-component-loading.md

```diff
@@ -0,0 +1,117 @@
+* XUnit loading tests - the test runner acts as an app and the test is loaded dynamically. The test can have any number of dependencies. Finding and resolving those dependencies is challenging.
+* ASP .NET's `dotnet watch` - ability to dynamically reload an app without restarting the process. Each version of the app is inherently in collision with any previous version. The old version should be unloaded.
+
+In lot of these cases the component which is to be loaded dynamically has a non-trivial set of dependencies which are unknown to the app itself. So the loading mechanism has to be able to resolve them.
+
+## Declaring dependencies
+The .NET Core SDK tooling produces `.deps.json` which is a dependency manifest for dotnet apps. It enables components (or apps) to load dependencies from locations other than the base directory (ex: from packages, platform-specific publish directories, etc.)  
+.NET Core first builds a set of directories to look for binaries (called ProbingPaths) based on application/component base directory, framework directory, `.runtimeconfig.dev.json`, command line, servicing locations, etc. For more details see [host-probing](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-probing.md).
+
+For an application or component, the `.deps.json` file specifies:
+* A set of binary dependencies -- which are searched within ProbingPaths
+* Relative paths to locate them -- relative paths in `.deps.json` file can be used to locate architecture-specific dependencies within publish/package directories.
+
+If the app depends on any frameworks, the `.deps.json` files of those framework are similarly processed.
+Further details about the algorithm used for processing dependencies can be found in [assembly-conflict-resolution](
+https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/assembly-conflict-resolution.md).
+
+## Dynamic loading with dependencies
+We propose to add a new public API which would dynamically load a component with these properties:
+* Component is loaded in isolation from the app (and other components) so that potential collisions are not an issue
+* Component can use `.deps.json` to describe its dependencies. This includes the ability to describe additional NuGet packages, RID-specific and/or native dependencies
+* Component can chose to rely on the app for certain dependencies by not including them in its `.deps.json`
+* Optionally such component can be enabled for unloading
+
+Public API (early thinking):
+```csharp
+class Assembly
+{
+    public static Assembly LoadFileWithDependencies(string path);
+}
+```
+
+In its core this is similar to `Assembly.LoadFile` but it supports resolving dependencies through `.deps.json` and such. Just like `Assembly.LoadFile` it provides isolation, but also for the dependencies.
+
+```csharp
+class AssemblyLoadContext
+{
+    public static AssemblyLoadContext CreateForAssemblyWithDependencies(string assemblyPath);
```

Comment:
I see little value for the plain `string` argument. We should have a binary path for this kind of thing either (a) the caller wants predictable and well-defined behavior or (b) the caller has some special scenario and they need to do some work. I think we are providing too many entry points for an already niche scenario in the grand scheme of things. We can _always_ add and make simpler, but removing is impossible.

Reply:
Makes sense